src/test/java/edu/hm/hafner/analysis/ast/CheckStyleAstFactoryTest.java

```diff
@@ -0,0 +1,795 @@
+        checkThatHashesMatching(NEED_BRACES, "NeedBraces10", REFACTORING_EXTRACT_CONSTANT, false);
+    }
+
+    /**
+     * Verifies that the ast calculates NOT the same hashcode.
+     */
+    @Test
+    void testPackageDeclarationWithExtractConstantHaveNotTheSameHashcode() {
+        checkThatHashesMatching(PACKAGE_DECLARATION, "PackageDeclaration4", REFACTORING_EXTRACT_CONSTANT, false);
+    }
+
+    /**
+     * Verifies that the ast calculates the same hashcode.
+     */
+    @Test
+    void testExplicitInitializationWithExtractConstant() {
+        checkThatHashesMatching(EXPLICIT_INITIALIZATION, "ExplicitInitialization5", REFACTORING_EXTRACT_CONSTANT, true);
+    }
+
+    /**
+     * Shows that a previous file has the same warnings like the current file, if refactorings were realised. It means
+     * that the same hashcode is calculated. (The warnings wasn't changed!)
+     */
+    @Test
+    void testFileWithManyWarningsHasSameHashcode() {
+        evaluateHashes("MoreWarningsInClass", "MoreWarningsInClass_Refactored", 5, 5, 5);
+    }
+
+    /**
+     * Shows that a previous file has more warnings than the current file, if warnings were fixed. Other new warnings
+     * were added.
+     */
+    @Test
+    void testPreviousHasMoreWarningsThanCurrent() {
+        evaluateHashes("MoreWarningsInClass2", "MoreWarningsInClass2_Refactored", 5, 4, 3);
+    }
+
+    /**
+     * Shows that a previous file has less warnings than the current file, if the current files have additional
+     * warnings. The difference in set theory is (current-warnings minus previous warnings), which are the new
+     * warnings.
+     */
+    @Test
+    void testCurrentHasMoreWarningsThanPrevious() {
+        evaluateHashes("MoreWarningsInClass", "MoreWarningsInClass1_Refactored", 5, 7, 5);
+    }
+
+    private void evaluateHashes(final String fileBefore, final String fileAfter, final int expectedPrev,
+            final int expectedCur, final int intersection) {
+        Set<String> hashSetPevious = calculateHashes(fileBefore, true);
+        Set<String> hashSetCurrent = calculateHashes(fileAfter, false);
+
+        HashSet<String> i = new HashSet<>(hashSetCurrent);
+        i.retainAll(hashSetPevious);
+
+        assertEquals("Not expected count of previous warnings", expectedPrev, hashSetPevious.size());
+        assertEquals("Not expected count of current warnings", expectedCur, hashSetCurrent.size());
+        assertEquals("The warnings aren't equal.", intersection, i.size());
+    }
+
+    private Set<String> calculateHashes(final String file, final boolean before) {
+        Report annotations;
+        String fileWithXmlExtension = file + ".xml";
+        String fileWithJavaExtension = file + ".java";
+
+        if (before) {
+            annotations = parse("before/" + fileWithXmlExtension);
+        }
+        else {
+            annotations = parse("after/" + fileWithXmlExtension);
+        }
+
+        Set<String> hashSet = new HashSet<String>();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Explicit type arguments can be replaced by Diamond Operator](https://app.codacy.com/manual/uhafner/analysis-model/pullRequest?prid=5019963)